portal/config/model_persistence.py

```diff
@@ -0,0 +1,237 @@
+
+from ..database import db
+from ..dict_tools import dict_match
+from ..date_tools import FHIR_datetime
+from ..models.identifier import Identifier
+
+
+class ModelPersistence(object):
+    """Adapter class to handle persistence of model tables"""
+    VERSION = '0.2'
+
+    def __init__(
+            self, model_class, lookup_field='id', sequence_name=None):
+        """Initialize adapter for given model class"""
+        self.model = model_class
+        self.lookup_field = lookup_field
+        self.sequence_name = sequence_name
+
+    def _log(self, msg):
+        current_app.logger.info(msg)
+
+    def __header__(self, data):
+        data['resourceType'] = 'Bundle'
+        data['id'] = 'SitePersistence v{}'.format(self.VERSION)
+        data['meta'] = {'fhir_comments': [
+            "export of dynamic site data from host",
+            "{}".format(current_app.config.get('SERVER_NAME'))],
+            'lastUpdated': FHIR_datetime.now()}
+        data['type'] = 'document'
+        return data
+
+    def __read__(self):
+        self.filename = persistence_filename(
+            scope=self.model.__name__)
+        with open(self.filename, 'r') as f:
+            data = json.load(f)
+        self.__verify_header__(data)
+        return data
+
+    def __write__(self, data, target_dir):
+        self.filename = persistence_filename(
+            scope=self.model.__name__,
+            target_dir=target_dir)
+        if data:
+            with open(self.filename, 'w') as f:
+                f.write(json.dumps(data, indent=2, sort_keys=True))
+            self._log("Wrote site persistence to `{}`".format(self.filename))
+
+    def __verify_header__(self, data):
+        """Make sure header conforms to what we're looking for"""
+        if data.get('resourceType') != 'Bundle':
+            raise ValueError("expected 'Bundle' resourceType not found")
+        if data.get('id') != 'SitePersistence v{}'.format(self.VERSION):
+            raise ValueError("unexpected SitePersistence version {}".format(
+                self.VERSION))
+
+    def export(self, target_dir):
+        d = self.__header__({})
+        d['entry'] = self.serialize()
+        self.__write__(data=d, target_dir=target_dir)
+
+    def import_(self, keep_unmentioned):
+        objs_seen = []
+        data = self.__read__()
+        for o in data['entry']:
+            if not o.get('resourceType') == self.model.__name__:
+                raise ValueError(
+                    "Import {} error, Found unexpected '{}' resource".format(
+                        self.model.__name__, o.get('resourceType')))
+            result = self.update(o)
+            db.session.commit()
+            if self.lookup_field == 'identifier':
+                objs_seen.append(result.id)
+            else:
+                objs_seen.append(getattr(result, self.lookup_field))
+
+        # Delete any not named
+        if not keep_unmentioned:
+            if self.lookup_field == 'identifier':
```

Comment:
I think you could remove the if/else block with a dict lookup (unless I missed something), something like this:

```
field = {'identifier':'id'}.get(self.lookup_field, self.lookup_field)
...

query = self.model.query.filter(~getattr(self.model, field).in_(objs_seen)) if objs_seen else []
```

Reply:
Different context. `self.lookup_field` is a ModelPersistence attribute, used to adapt different model classes - some are looked up by id, some name, some use the related 'identifier' table.